### PR TITLE
Remove unused FSDP components

### PR DESCRIPTION
**Pull request by ebsmothers**
I was attempting to patch the changes from #1933 so we can merge them, but I'm unable to push to the remote branch (likely because the PR was opened off of the fork's main branch which has not yet been merged with latest changes from upstream). So I've just patched everything into this new PR instead.

Tagging @krammnic as the author of the original PR. The only changes I've made were merging latest changes from main and updating `test_validate_missing_and_unexpected_for_lora` to get it to pass (and adding some of the test cases from `validate_state_dict_for_lora` that were deleted)

### CODE REVIEW

**Comment by SalmanMohammadi** [comment]
One small thing - is it worth deprecating `validate_state_dict_for_lora` and deleting it next release, since it's a public API? For the hackers.

**Review by SalmanMohammadi — APPROVED** [review]

File: docs/source/api_ref_modules.rst [hunk 1]
```diff
@@ -77,7 +77,6 @@ PEFT Components
     peft.set_trainable_params
     peft.get_adapter_state_dict
     peft.validate_missing_and_unexpected_for_lora
-    peft.validate_state_dict_for_lora
```
Comment (SalmanMohammadi) [inline docs/source/api_ref_modules.rst:80]: Line 209 in the LoRA finetune tutorial:
> .. note::
    Whenever loading weights with :code:`strict=False`, you should verify that any missing or extra keys in
    the loaded :code:`state_dict` are as expected. torchtune's LoRA recipes do this by default via e.g.
    :func:`validate_state_dict_for_lora() <torchtune.modules.peft.validate_state_dict_for_lora>` or
    :func:`validate_missing_and_unexpected_for_lora() <torchtune.modules.peft.validate_missing_and_unexpected_for_lora>`.

Needs to be updated

File: docs/source/api_ref_training.rst [hunk 2]
```diff
@@ -50,12 +50,9 @@ Utilities for enabling and working with distributed training.
     :toctree: generated/
     :nosignatures:
 
-    FSDPPolicyType
     init_distributed
     is_distributed
     get_world_size_and_rank
-    get_full_finetune_fsdp_wrap_policy
```
Comment (SalmanMohammadi) [inline docs/source/api_ref_training.rst:57]: There's a reference to this in the [QAT tutorial](https://github.com/pytorch/torchtune/blob/ac14e96443e60fd8ffed434637357ce22a9c5bff/docs/source/tutorials/qat_finetune.rst?plain=1#L173) too.

**Comment by ebsmothers** [comment]
> One small thing - is it worth deprecating the public APIs and deleting next release? I'm not sure if people are really using these so your call

Yeah this is a fair question. In this case I am gonna yolo a bit because we haven't supported FSDP1 for multiple releases at this point. And I think most of these APIs should not really have been public to begin with, but we made them so by necessity because many were used directly in our recipes

**Pull request merged**